src/mongo/db/storage/csv_file.cpp

```diff
@@ -441,28 +458,88 @@ boost::optional<BSONObj> CsvFileInput::readBsonObj() {
     return builder.done().getOwned();
 }
 
+/**  State Machine Diagram:
+ * Start State: ⬇︎  ︎         cur_char != "
+ * +--------------+  The field is NOT enclosed    +----------------+
+ * |    State:    | ----------------------------> |     State:     |-----
+ * |  begindField |                               |  quote_Closed  |     | regular characters
+ * |              | <---------------------------- |                | <---
+ * +--------------+ <___            cur_char = ,  +----------------+ ---> cur_char = " illegal
+ *   |  ⬆︎               \             End of field               double quote, RFC Non compliant,
+ *   |__| cur_char = ,    \_____
+ *   |    Empty field           |___________
+ *   |                                      \
+ *   |                                       \
+ *   | cur_char = "                           ︎\ ︎ ︎  cur_char = ,
+ *   | The field is enclosed by                \   Meaning next character after closing quote is
+ *   | Double Quotes                            \  comma, (",) Correct syntax for end of field
+ *   ⬇︎ ︎                                          \
+ * +--------------+       cur_char = "            +----------------+
+ * |     State:   | ----------------------------> |      State:    | other characters
+ * |  quote_open  |                               | checkForDouble | ---> RFC Non compliant
+ * |              | <---------------------------- |     -Quote     |
+ * +--------------+       cur_char = "            +----------------+
+ *    |   ⬆︎          Double double quote (""),
+ *    |   |           Escaped double quote
+ *    |___|
+ *   Any character, except double quote "
+ */
+
 std::vector<std::string_view> CsvFileInput::parseLine(const std::string_view& line) {
-    int left = 0;
-    int right = 0;
-    int len = line.length();
-    int i = 0;
+    auto stateMachine = parsingState::beginField;
+
+    const size_t len = line.length();
+    size_t i = 0;
+    size_t left = 0;
     std::vector<std::string_view> fields;
 
     while (i <= len) {
-        if (i == len || line.at(i) == ',') {
-            fields.emplace_back(line.substr(left, right - left));
-            left = 0;
-            right = 0;
-        } else if (line.at(i) != ' ' && line.at(i) != '\t') {
-            if (right == 0) {
+        switch (stateMachine) {
+            case parsingState::beginField:
                 left = i;
-            }
-            right = i + 1;
+                if (i == len || line[i] == ',') {  // Empty field.
+                    fields.emplace_back(line.substr(left, 0));
+                    left++;
+                    break;
+                }
+                stateMachine =
+                    line[i] == '\"' ? parsingState::quoteOpen : parsingState::quoteClosed;
+                break;
+            case parsingState::quoteOpen:
+                if (i == len) {  // Field that does not comply with RFC 4180
```

Comment:
[nit] The comment is repeating _ioStats->incNonCompliantWithRFC(). Let's add more useful comment, not repeating the code.

```suggestion
                if (i == len) {  // Reached the end of record without closing the previous double quote.
```